### PR TITLE
feat: add translations to public form

### PR DESCRIPTION
**Pull request by siddarth2824**
## Changes
- Add language control to allow public form users to toggle languages on the form
- Use admin input translations on public form
- Add translations for prevent submission from form logic
- Add fixed translations for some components on the public form which are not generated by admin. For example, labels in `FormInstructionsContainer`. This [sheet](https://docs.google.com/spreadsheets/d/1u_dH53sglH5w7jV83rU5NFIhjk80_FIJ/edit?usp=sharing&ouid=101878657257239680436&rtpof=true&sd=true) contains the fixed translations. Highlighted in green are the translations added in this PR.

## Before & After Screenshots
| Description | Before | After | 
| ----------- | ------- | ----- |
| Language control component to allow users to select the language to view the public form in | | Screenshot 2024-12-11 at 5 24 24 PM |

### CODE REVIEW

**Comment by LoneRifle** [comment]
Generally, we would use i18next for text. Would be happy to point out how wthat can be done.

**Review by LoneRifle — DISMISSED** [review]
Yea I'm afraid I might have to ask you to refactor to incorporate the use of i18next, as we discussed the other time

File: shared/constants/field/myinfo/index.ts [hunk 1]
```diff
@@ -50,6 +53,11 @@ export const types: MyInfoFieldBlock[] = [
       'The registered name of the form-filler. This field is verified by ICA for Singaporeans/PRs & foreigners on Long-Term Visit Pass, and by MOM for Employment Pass holders.',
     fieldType: BasicField.ShortText,
     previewValue: 'PHUA CHU KANG',
+    titleTranslations: [
+      { language: Language.CHINESE, translation: '姓名' },
+      { language: Language.MALAY, translation: 'Nama' },
+      { language: Language.TAMIL, translation: 'பெயர்' },
+    ],
```
Comment (LoneRifle) [inline shared/constants/field/myinfo/index.ts:60]: Er we've talked about this, can you use i18next please?

File: frontend/src/features/admin-form/create/builder-and-design/utils/fieldCreation.ts [hunk 2]
```diff
@@ -220,11 +220,19 @@ export const getMyInfoFieldCreationMeta = (
...
+    | 'description'
+    | 'fieldType'
+    | 'myInfo'
+    | 'titleTranslations'
+    | 'descriptionTranslations'
   > = {
     disabled: false,
     required: true,
     title: MYINFO_ATTRIBUTE_MAP[myInfoAttribute].value,
+    titleTranslations: MYINFO_ATTRIBUTE_MAP[myInfoAttribute]?.titleTranslations,
```
Comment (KenLSM) [inline frontend/src/features/admin-form/create/builder-and-design/utils/fieldCreation.ts:235]: Note to self: this is expected to remains as a temporary solution as it is complex to extract the translations to be specified in locales file.

File: frontend/src/features/admin-form/settings/SettingsPage.tsx [hunk 3]
```diff
@@ -51,6 +56,18 @@ export const SettingsPage = (): JSX.Element => {
...
+    isUserLoading ||
+    isFormSettingLoading ||
+    !user?.betaFlags?.multiLangTranslation
+      ? null
+      : {
+          label: 'Multi-language',
+          icon: LanguageTranslation,
+          component: SettingsMultiLangPage,
+          path: 'language',
+        }
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/SettingsPage.tsx]: /nit it is easier to read the inverse.

```suggestion
    !isUserLoading &&
    !isFormSettingLoading &&
    user?.betaFlags?.multiLangTranslation
      ? {
          label: 'Multi-language',
          icon: LanguageTranslation,
          component: SettingsMultiLangPage,
          path: 'language',
        }
      : null
```

File: shared/constants/field/myinfo/index.ts
Comment (KenLSM) [inline shared/constants/field/myinfo/index.ts:1]: Note to self: sister file relation to translations within a constants file as interim instead of having in locales.

File: frontend/src/features/public-form/components/FormFields/PublicFormSubmitButton.tsx [hunk 4]
```diff
@@ -65,6 +65,19 @@ export const PublicFormSubmitButton = ({
     })
   }, [formInputs, formFields, formLogics])
 
+  const selectedLanguage = i18n.language as Language
+  const preventSubmissionMessage = useMemo(() => {
```
Comment (KenLSM) [inline frontend/src/features/public-form/components/FormFields/PublicFormSubmitButton.tsx]: Possible to use `getValueInSelectedLanguage()`?

File: frontend/src/templates/Field/Radio/RadioField.tsx [hunk 5]
```diff
@@ -50,6 +53,8 @@ export const RadioField = ({
     [disableRequiredValidation, schema],
   )
 
+  const defaultEnglishRadioOptions = schema.fieldOptions
```
Comment (KenLSM) [inline frontend/src/templates/Field/Radio/RadioField.tsx]: /supernit would suggest to not name it as `defaultEnglishRadioOptions` but rather straight up `englishRadioOptions`. 

I had initially through that this value is to be overwritten as `defaultX` suggests that the value would be replaced. However, this is not the case as we strictly only want to pass `englishRadioOptions` value to the backend.

File: frontend/src/templates/Field/Dropdown/DropdownField.tsx [hunk 6]
```diff
@@ -23,12 +25,53 @@ export const DropdownField = ({
...
   const rules = useMemo(() => {
     return createDropdownValidationRules(schema, disableRequiredValidation)
   }, [schema, disableRequiredValidation])
 
   const { control } = useFormContext<SingleAnswerFieldInput>()
 
+  const selectedLanguage = i18n.language as Language
+
+  const fieldOptions: ComboboxItem[] = useMemo(() => {
+    const defaultEnglishFieldOptions = schema.fieldOptions
```
Comment (KenLSM) [inline frontend/src/templates/Field/Dropdown/DropdownField.tsx]: /supernit in similar style with `RadioField`. `defaultX` suggests that this value would be replaced, but we shouldn't since we only want to pass the english values to the backend.

File: frontend/src/templates/Field/Dropdown/DropdownField.tsx [hunk 7]
```diff
@@ -23,12 +25,53 @@ export const DropdownField = ({
...
+    const translationIdx = fieldOptionsTranslations.findIndex((translation) => {
+      return translation.language === selectedLanguage
+    })
+
+    // Check if translations for field options exist and whether
+    // each field option has its own respective translation. If not
+    // render the default field options in English.
+    if (
+      translationIdx !== -1 &&
+      fieldOptionsTranslations[translationIdx].translation.length ===
```
Comment (KenLSM) [inline frontend/src/templates/Field/Dropdown/DropdownField.tsx:51]: Check why this change is much more than radio field

File: frontend/src/templates/Field/Checkbox/CheckboxField.tsx [hunk 8]
```diff
@@ -59,11 +62,20 @@ export const CheckboxField = ({
     [disableRequiredValidation, schema],
   )
 
+  const defaultEnglishCheckboxOptions = schema.fieldOptions
```
Comment (KenLSM) [inline frontend/src/templates/Field/Checkbox/CheckboxField.tsx]: /supernit in similar style with `RadioField`. `defaultX` suggests that this value would be replaced, but we shouldn't since we only want to pass the english values to the backend.

File: frontend/src/utils/multiLanguage.ts [hunk 9]
```diff
@@ -16,3 +20,64 @@ export const convertUnicodeLocaleToLanguage = (language: Language) => {
...
+  // Check if translations for field options exist and whether
+  // each field option has its own respective translation. If not
+  // render the default field options in English.
+  if (
+    translationIdx !== -1 &&
+    fieldOptionsTranslations[translationIdx].translation.length ===
+      defaultValue.length
+  ) {
+    return fieldOptionsTranslations[translationIdx].translation
+  } else {
```
Comment (KenLSM) [inline frontend/src/utils/multiLanguage.ts]: /nit this `else` can be omitted. I generally prefer an `if without else` aka guard clauses

File: frontend/src/utils/multiLanguage.ts [hunk 10]
```diff
@@ -16,3 +20,64 @@ export const convertUnicodeLocaleToLanguage = (language: Language) => {
...
+  translations?: TranslationOptionMapping[]
+  selectedLanguage: Language
+}
+
+export const getValueInSelectedLanguage = ({
+  defaultValue,
+  translations,
+  selectedLanguage,
+}: SelectedLanguageProps) => {
+  let title = defaultValue
```
Comment (KenLSM) [inline frontend/src/utils/multiLanguage.ts]: /supernit Why `title`? This function is trying to get the `value` but this suggest that the value is used for `titles`

File: frontend/src/templates/Field/Dropdown/DropdownField.tsx (hunk 7 above)
Comment (KenLSM) [reply frontend/src/templates/Field/Dropdown/DropdownField.tsx:51]: Is the reason why we cannot use `getFieldOptionsInSelectedLanguage` is due to the fact that we have a custom extractor[1] here? 

Context of question: I do like the utils function as it abstracts out the selection language out of the component rendering logic, so I'm sadden to see that it cannot be utilised. When one is reading through this component, the complex logic of language selection / detection (such as detection and selection of language + its defaults) is a distraction.

[1] the assumed offending extractor
```
return translatedFieldOptions.map((translatedFieldOption, index) => {
  return {
    value: defaultEnglishFieldOptions[index],
    label: translatedFieldOption,
  }
})
```

If this is the only problem, we can likely look into a utils that supports custom extractors as an enhancement in the future.

File: frontend/src/templates/Field/Section/SectionField.tsx [hunk 11]
```diff
@@ -39,6 +41,34 @@ export const BaseSectionField = forwardRef<
     },
   })
 
+  const selectedLanguage = i18n.language as Language
+
+  let title = schema.title
+  const titleTranslations = schema?.titleTranslations ?? []
```
Comment (KenLSM) [inline frontend/src/templates/Field/Section/SectionField.tsx]: Can use `getValueInSelectedLanguage`?

File: frontend/src/templates/Field/Section/SectionField.tsx [hunk 12]
```diff
@@ -39,6 +41,34 @@ export const BaseSectionField = forwardRef<
...
+  })
+
+  // If there exists a translation for title based on the selected language,
+  // use that. If not default to english
+  if (titleTranslationIdx !== -1) {
+    title = titleTranslations[titleTranslationIdx].translation
+  }
+
+  let description = schema.description
+  const descriptionTranslations = schema?.descriptionTranslations ?? []
```
Comment (KenLSM) [inline frontend/src/templates/Field/Section/SectionField.tsx]: Can use `getValueInSelectedLanguage`?

File: frontend/src/templates/Field/Table/ColumnCell.tsx [hunk 13]
```diff
@@ -82,12 +84,52 @@ const DropdownColumnCell = ({
...
+    // Check if translations for field options exist and whether
+    // each field option has its own respective translation. If not
+    // render the default field options in English.
+    if (
+      translationIdx !== -1 &&
+      fieldOptionsTranslations[translationIdx].translation.length ===
+        defaultEnglishFieldOptions.length
+    ) {
+      const translatedFieldOptions =
+        fieldOptionsTranslations[translationIdx].translation
```
Comment (KenLSM) [inline frontend/src/templates/Field/Table/ColumnCell.tsx:113]: Same question as in `frontend/src/templates/Field/Dropdown/DropdownField.tsx` regarding the utilisation of multi-lang utils. Was it due to a need for custom extractor?

File: frontend/src/templates/Field/Table/TableFieldContainer.tsx [hunk 14]
```diff
@@ -21,8 +24,42 @@ export const TableFieldContainer = ({
...
+    },
+  )
+
+  // If there are title translations for the selected language, use the translation.
+  // If not default it to English.
+  if (titleTranslationIdx !== -1) {
+    title = titleTranslations[titleTranslationIdx].translation
+  }
+
+  let description = schema.description
```
Comment (KenLSM) [inline frontend/src/templates/Field/Table/TableFieldContainer.tsx]: Can use `getValueInSelectedLanguage`?

File: frontend/src/features/public-form/components/LanguageControl/LanguageControl.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+  if (!shouldShowLanguageDropdown) return null
+
+  return (
+    <Flex
+      background={bgColour}
+      zIndex={10}
+      px={{ base: '1.5rem', md: 0 }}
+      justifyContent={{ base: 'start', md: 'center' }}
+    >
+      <HStack mt="-32px" bg="white" borderRadius="4px" shadow="md">
```
Comment (KenLSM) [inline frontend/src/features/public-form/components/LanguageControl/LanguageControl.tsx]: Able to use `rem`?

File: frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx [hunk 16]
```diff
@@ -345,9 +375,26 @@ export const TranslationListSection = ({
             )
           })}
 
+          {/* Form Logic Prevent Submission Message Translation */}
+          {formLogicPreventSubmissions && (
+            <>
+              <QuestionRow
+                key="Form Logic"
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx:382]: Is `key` needed here since we're not rendering them in a list like manner.

File: frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx [hunk 17]
```diff
@@ -345,9 +375,26 @@ export const TranslationListSection = ({
...
+                unicodeLocale={language}
+              />
+              <Divider />
+            </>
+          )}
+
           {/* End Page Translation */}
           {form.endPage && (
             <QuestionRow
+              key="End Page"
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx:397]: Is `key` needed here since we're not rendering them in a list like manner.

File: frontend/src/components/FormEndPage/EndPageBlock.tsx [hunk 18]
```diff
@@ -43,6 +46,31 @@ export const EndPageBlock = ({
...
 
+  const selectedLanguage = i18n.language as Language
+
+  const title = getValueInSelectedLanguage({
+    defaultValue: endPage.title,
+    translations: endPage.titleTranslations,
+    selectedLanguage,
+  })
+
+  const paragraph = useMemo(() => {
```
Comment (KenLSM) [inline frontend/src/components/FormEndPage/EndPageBlock.tsx]: Can use `getValueInSelectedLanguage`?

**Review by KenLSM — COMMENTED** [review]
Several question. All non-blocking.

File: frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx (hunk 16 above)
Comment (siddarth2824) [reply frontend/src/features/admin-form/settings/components/MultiLanguageSection/TranslationListSection.tsx:382]: Even though we are not rendering in a list manner, I think its required because there are other `<QuestionRow>` components being rendered at the same nested level as itself. Hence requiring a key to uniquely identify the `<QuestionRow>` components from each other. I keep seeing the error on my console regarding this, hence I added it. I am not very sure if its required, but when I add it, the error on console seems to disappear.

**Review by KenLSM — APPROVED** [review]
LGTM! Thanks for taking time to split the PRs for easier review 👍
